addons/binding/org.openhab.binding.modbus/DEVELOPERS.md

```diff
@@ -0,0 +1,111 @@
+# For Developers
+
+## Testing Serial Implementation
+
+You can use test serial slaves without any hardware on linux using these steps:
+
+1. Set-up virtual null modem emulator using [tty0tty](https://github.com/freemed/tty0tty)
+2. Download [diagslave](http://www.modbusdriver.com/diagslave.html) and start modbus serial slave up using this command:
+
+```
+./diagslave -m rtu -a 1 -b 38400 -d 8 -s 1 -p none -4 10 /dev/pts/7
+```
+
+3. Configure openHAB's modbus slave to connect to `/dev/pts/8`:
+
+```
+xxx.connection=/dev/pts/8:38400:8:none:1:rtu
+```
+
+4. Modify `start.sh` or `start_debug.sh` to include the unconventional port name by adding the following argument to `java`:
+
+```
+-Dgnu.io.rxtx.SerialPorts=/dev/pts/8
+```
+
+Naturally this is not the same thing as the real thing but helps to identify simple issues.
+
+## Testing TCP Implementation
+
+1. Download [diagslave](http://www.modbusdriver.com/diagslave.html) and start modbus tcp server (slave) using this command:
+
+```
+./diagslave -m tcp -a 1 -p 55502
+```
+
+2. Configure openHAB's modbus slave to connect to `127.0.0.1:55502`:
+
+```
+tcp.slave1.connection=127.0.0.1:55502
+```
+
+## Writing Data
+
+See this [community post](https://community.openhab.org/t/something-is-rounding-my-float-values-in-sitemap/13704/32?u=ssalonen) explaining how `pollmb` and `diagslave` can be used to debug modbus communication.
+
+You can also use `modpoll` to write data:
+
+
+```bash
+# write value=5 to holding register 40001 (index=0 in the binding)
+./modpoll -m tcp -a 1 -r 1 -t4 -p 502 127.0.0.1 5
+# set coil 00001 (index=0 in the binding) to TRUE
+./modpoll -m tcp -a 1 -r 1 -t0 -p 502 127.0.0.1 1
+# write float32
+./modpoll -m tcp -a 1 -r 1 -t4:float -p 502 127.0.0.1 3.14
+```
+
+## Extending Modbus binding
```

Comment:
It's good that you moved "For developers" to separate place.

However, it would make sense to mention the existence of device specific bindings in PR https://github.com/openhab/openhab2-addons/pull/4220 . We would like to follow the example set by bluetooth docs: https://www.openhab.org/addons/bindings/bluetooth/#bluetooth-binding "It has the following extensions:", or in [markdown](https://raw.githubusercontent.com/openhab/openhab2-addons/master/addons/binding/org.openhab.binding.bluetooth/README.md)

```

It has the following extensions:

<!--list-subs-->
```

Reply:
Ok, this seems reasonable. Once this PR gets merged, I will rebase that PR completely, and add update the documentation as well.
But I don't see any point of adding an empty section right now.

Reply:
Thanks, agreed.